memtagabielf64/memtagabielf64.rst

```diff
@@ -0,0 +1,723 @@
+non-MemtagABI case, and in cases where the relocation operation is utilised for
+an untagged symbol, the target field will be zero, and so the tag derivation
+will come directly from ``*(Delta(S) + A)``, maintaining backwards
+compatibility. This has a few effects though:
+
+1. MemtagABI binaries must use RELA relocation encoding, as the unused bits in
+   the target field are used as metadata in the relocation operation.
+
+2. Out-of-bounds relocations to tagged symbols must use the
+   ``R_AARCH64_RELATIVE`` relocation type when ``ABS64`` or ``GLOB_DAT``
+   relocations are not available. The tag derivation offset must be encoded in
+   the target field. In these instances, because the target field is now no
+   longer zero, some global variables may move from the zero-initialized ``bss``
+   section to the constant-initialized ``data`` section.
+
+3. ``RELR``-style relocation compression can only be applied to
+   ``R_AARCH64_RELATIVE`` relocations that have a zero tag-derivation offset
+   (i.e. the tag derivation can be done entirely by ``ldg`` of the
+   tag-granule-aligned result of the relocation operation).
+
+An alternative considered was to use an ``ABS64`` or ``GLOB_DAT`` relocation
+instead, to avoid having to materialize the tag-derivation offset. Usinhg these
+relocations would however require additional dynamic symbol table entries for
+each symbol used, polluting the dynsym table unnecessarily for each tagged
+internal symbol.
+
+.. raw:: pdf
+
+   PageBreak
+
+Run-time dynamic linking
+========================
+
+Process initialization using dynamic table entries
+--------------------------------------------------
+
+During process intialization, the dyanmic loader should use the MemtagABI
+dynamic table entries to enable the correct MTE features at runtime.
+
+If ``DT_AARCH64_MEMTAG_MODE`` is present, the dynamic loader should (in a
+platform-specific specific way) enable MTE for the process. On Linux, this is
+done through ``prctl(PR_SET_TAGGED_ADDR_CTRL, PR_TAGGED_ADDR_ENABLE |
+PR_MTE_TCF_[A]SYNC ...)``. When ``DT_AARCH64_MEMTAG_MODE`` is ``0``, then
+synchronous MTE should be enabled. When ``DT_AARCH64_MEMTAG_MODE`` is ``1``,
+asynchronous MTE should be enabled.
+
+If ``DT_AARCH64_MEMTAG_HEAP`` is present, the dynamic loader should inform the
+heap allocator (in a platform-specific way) that memory tags should be used for
+all heap allocations in this process.
+
+If ``DT_AARCH64_MEMTAG_STACK`` is present, the dynamic loader should enable
+tagging for the main stack and thread stacks. Re-mapping currently alive stack
+frames as taggable is dangerous, and so dynamic loader developers may find it
+easiest to implement this functionality by:
+
+1. Ensuring the implementation completes before any threads have spawned, and
+
+2. Creating a new MTE-protected main stack and calling a function that
+   transitions onto the new stack, preserving all existing frames on the
+   untagged stack.
+
+Global tagging
```

Comment:
Should this be "Globals tagging" (plural)? Or "Global variable tagging", which is a bit less awkward.